api/workload/v2/workload.go

```diff
@@ -0,0 +1,235 @@
+type X509SVIDClient struct {
+	watcher      WorkloadX509SVIDWatcher
+	addr         string
+	wg           sync.WaitGroup
+	ctx          context.Context
+	cancelFn     func()
+	backoff      *backoff
+	stateManager *clientStateManager
+}
+
+// Option configures the workload client.
+type Option func(*X509SVIDClient)
+
+// WithAddr specifies the unix socket address of the SPIFFE agent.
+func WithAddr(addr string) Option {
+	return func(w *X509SVIDClient) {
+		w.addr = addr
+	}
+}
+
+// NewX509SVIDClient returns a new Workload API client for X.509 SVIDs.
+func NewX509SVIDClient(watcher WorkloadX509SVIDWatcher, opts ...Option) (*X509SVIDClient, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &X509SVIDClient{
+		addr:         DefaultAgentAddress,
+		watcher:      watcher,
+		ctx:          ctx,
+		cancelFn:     cancel,
+		backoff:      newBackoff(),
+		stateManager: newClientStateManager(),
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	if !strings.HasPrefix(c.addr, _unixPathPrefix) {
+		return nil, fmt.Errorf("spiffe/workload: agent address %q is not a unix address", c.addr)
+	}
+	return c, nil
+}
+
+// Start starts the client.
+//
+// The client will always start, and users should rely on the watcher
+// interface to receives updates on the client's status.
+//
+// It is an error to call Start() more than once. Calling Start() after
+// Stop() is not supported.
+func (c *X509SVIDClient) Start(ctx context.Context) error {
+	if err := c.stateManager.StartIfStartable(); err != nil {
+		return fmt.Errorf("spiffe/workload: %v", err)
+	}
+	c.wg.Add(1)
+	go c.run()
+	return nil
+}
+
+func (c *X509SVIDClient) run() {
+	defer c.wg.Done()
+
+	conn := c.newConn()
+	if conn == nil {
+		return
+	}
+	c.watcher.OnConnection(true)
+	defer c.watcher.OnConnection(false)
+	defer conn.Close()
+
+	c.watch(conn)
+}
+
+// establishes a new persistent connection, returns nil if a connection can't be created
+func (c *X509SVIDClient) newConn() *grpc.ClientConn {
+	for {
+		conn, err := grpc.DialContext(c.ctx, c.addr, grpc.WithInsecure())
+		if err != nil {
+			if done := c.handleError(err); done {
+				return nil
+			}
+			continue
+		}
+		c.backoff.Reset()
+		return conn
+	}
+}
+
+// handles an error, applies backoff, and returns true if the context has been canceled
+func (c *X509SVIDClient) handleError(err error) (done bool) {
+	if status.Code(err) == codes.Canceled {
+		return true
+	}
+	c.watcher.OnError(err)
+	select {
+	case <-time.After(c.backoff.Duration()):
+		return false
+	case <-c.ctx.Done():
+		return true
+	}
+}
+
+// creates single watch loop for the connection
+func (c *X509SVIDClient) watch(conn *grpc.ClientConn) {
+	for {
+		ctx, cancel := context.WithCancel(c.ctx)
+		stream, err := c.newX509SVIDStream(ctx, conn)
+		if err != nil {
+			cancel()
+			if done := c.handleError(err); done {
+				return
+			}
+			continue
+		}
+		if err := c.handleX509SVIDStream(stream); err != nil {
```

Comment:
right now handleX509SVIDStream always returns an error. However, if a future refactor changes that, the context you created above will not be cancelled and we'll leak contexts. Maybe something like:

```
func watch(conn) {
    for {
        if done := watchOnce(conn); done {
            return true
        }
    }
}

func watchOnce(conn) {
    ctx, cancel := ...
    defer cancel()

    // rest of the body
}
```